### PR TITLE
Use pathlib for edit magic command

### PR DESCRIPTION
**Pull request by rchiodo**
Change 'with open' to use Pathlib instead.
See https://github.com/ipython/ipython/issues/12515.

### CODE REVIEW

File: IPython/core/magics/code.py [hunk 1]
```diff
@@ -512,7 +513,7 @@ def _edit_macro(self,mname,macro):
         self.shell.hooks.editor(filename)
 
         # and make a new macro object, to replace the old one
-        with open(filename) as mfile:
+        with Path(filename).open() as mfile:
             mvalue = mfile.read()
```
Comment (ashwinvis) [inline IPython/core/magics/code.py:517]:
```suggestion
        mvalue = Path(filename).read_text()
```

File: IPython/core/magics/code.py [hunk 2]
```diff
@@ -688,19 +689,22 @@ def edit(self, parameter_s='',last_call=['','']):
...
         except TryNext:
             warn('Could not open editor')
             return
 
         # XXX TODO: should this be generalized for all string vars?
         # For now, this is special-cased to blocks created by cpaste
         if args.strip() == 'pasted_block':
-            with open(filename, 'r') as f:
+            with filepath.open('r') as f:
                 self.shell.user_ns['pasted_block'] = f.read()
```
Comment (ashwinvis) [inline IPython/core/magics/code.py:708]:
```suggestion
            self.shell.user_ns['pasted_block'] = filepath.read_text()
```

File: IPython/core/magics/code.py [hunk 3]
```diff
@@ -711,7 +715,7 @@ def edit(self, parameter_s='',last_call=['','']):
                 if not is_temp:
                     self.shell.user_ns['__file__'] = filename
                 if 'r' in opts:    # Untranslated IPython code
-                    with open(filename, 'r') as f:
+                    with filepath.open('r') as f:
                         source = f.read()
```
Comment (ashwinvis) [inline IPython/core/magics/code.py:719]: Same here?

File: IPython/core/magics/code.py [hunk 4]
```diff
@@ -720,10 +724,10 @@ def edit(self, parameter_s='',last_call=['','']):
 
         if is_temp:
             try:
-                with open(filename) as f:
+                with filepath.open() as f:
                     return f.read()
```
Comment (ashwinvis) [inline IPython/core/magics/code.py:728]: And here?

**Review by ashwinvis — CHANGES_REQUESTED** [review]
`Path.read_text` method can be used to condense this code further. Should work on Python >= 3.5

**Comment by Carreau** [comment]
Good teamwork !

I've push some formatting changes and will merge once tests are passing.

**Pull request merged**